File: JavaScript/713-Subarray Product Less Than K.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var numSubarrayProductLessThanK = function(nums, k) {
    
    if (k <= 1){
        return 0;
    }
    
    let res = 0, left = 0, product = 1;
    
    for (let i=0; i<nums.length; i++){
        product *= nums[i];
        while (product >= k){
            product /= nums[left];
            left ++;
        }
        res += i - left + 1;
    }
    return res;
};

File: C++/713-Subarray Product Less Than K.cpp
class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
        int current = 1, left = 0, res = 0;
        for (int i=0; i<nums.size(); i++) {
            current *= nums[i];
            while (left <= i && current >= k) {
                current = current/nums[left];
                left++;
            }
            res += i-left+1;
        }
        
        return res;
    }
};